container relationships view changes

moolitayer opened the pull request:
After this change the following must hold for all container entity screens:
- both the relationships table and the navbar on the left must maintain the following order:
 provider, project, route, service, replicator, pod, container, node[1]
- ui tabs must maintain the same order
- navbar relationships must include all entities in table relationships
- all entities in relationships must be clickable
- if an entity has a relationship with a single parent (a pod belongs to one project), the navbar will show:
singular entity name: name
- if an entity has a one to many relationship (a node can have many pods), the navbar will show:
plural entity name: (#count)
- names of nav bar entities should be acquired using ui_lookup() and should not be translatable
- titles of the above should be translatable, and should always begins with 'show' and not 'view'

In addition:
- Use loops where possible in navbar code
- Simplify list view code where possible 
- Fixes #4893
- Fixes #4950

[1] This list does not include images and image registries;
at the images screen, image registry is listed second after provider
at the images registry screen, images are listed second after provider
at the provider screen, registries and then images are listed at the bottom

moolitayer commented:
@miq-bot add_label wip, providers/containers, ui, bug

moolitayer commented:
todo: tabs should probably have the same order described in https://github.com/ManageIQ/manageiq/pull/5029#issue-112633940 
So for example 'nodes' tab should move

abonas commented:
@moolitayer I think the PR's title is misleading. it says "Container entity view changes" - which is one specific entity type, while iiuc you are focusing here on all entities of the container providers.

moolitayer commented:
@miq-bot remove_label wip
@miq-bot add_label refactoring

moolitayer commented:
@simon3z @abonas please review
@dclarizio @martinpovolny @h-kataria please review

providers:
![manageiq containers providers](https://cloud.githubusercontent.com/assets/3010449/10667907/7f74174e-78e2-11e5-8426-96097a0d5222.png)

projects:
![manageiq projects](https://cloud.githubusercontent.com/assets/3010449/10667908/7f744156-78e2-11e5-8720-01846ca64741.png)

routes:
![manageiq routes](https://cloud.githubusercontent.com/assets/3010449/10667909/7f748fbc-78e2-11e5-82a5-35c341d7ca2a.png)

services:
![manageiq services](https://cloud.githubusercontent.com/assets/3010449/10667911/7f750fb4-78e2-11e5-859a-32e462bbb691.png)

replicators:
![manageiq replicators](https://cloud.githubusercontent.com/assets/3010449/10667910/7f74e7dc-78e2-11e5-9731-8bd51715d673.png)

pods:
![manageiq pods](https://cloud.githubusercontent.com/assets/3010449/10667912/7f76f2c0-78e2-11e5-8eae-adfd336c32f4.png)

nodes:
![manageiq nodes](https://cloud.githubusercontent.com/assets/3010449/10667914/7f9664d4-78e2-11e5-82c4-8d551de9bbef.png)

image registries:
![manageiq image registries](https://cloud.githubusercontent.com/assets/3010449/10667913/7f9259d4-78e2-11e5-8613-37bed7da401a.png)

images:
![manageiq images](https://cloud.githubusercontent.com/assets/3010449/10667915/7f96f7fa-78e2-11e5-8869-3ce4368b52b6.png)

moolitayer commented:
addressed relevant rubocop comments

zeari commented on app/helpers/application_helper.rb:
```diff
@@ -24,6 +24,42 @@ def miq_accordion_panel(title, condition, id, &block)
+    end
+    out
+  end
+
```
I'm guessing these functions should be in somewhere like ```listnav_helper.rb``` the same way ```textual_link``` is in ```textual_summary_helper.rb```
@matthewd ?

moolitayer replied:
@matthewd what would you like to see here?

blomquisg replied:
Hrm, I'm not sure `single_relationship_link` is gonna work correctly ... correct me if I'm missing something.

This block from `app/views/layouts/listnav/_container_group.html.haml` was replaced with a call to `single_relationship_link`:

> ```ruby
>      - if role_allows(:feature => "container_node") && !@record.container_node.nil?
>        %li
>          = link_to("#{ui_lookup(:table => "container_node")}: #{@record.container_node.name}",
>              {:controller => 'container_node', :action => 'show', :id => @record.container_node},
>              :title => _("View %s %s") % [ui_lookup(:table => "container_node"), @record.container_node.name])
> ```

And the corresponding call to `single_relationship_link`:

> ```ruby
>  = single_relationship_link(@record, :container_node)
> ```

Notice that in the original code, the call to `role_allows` does not include `"_show"` in the feature name.  However the implementation of `single_relationship_link` will always append `"_show"` to the feature name.

I think you'll have to include an optional "feature_name" to the arguments for this method, probably.

moolitayer replied:
nice catch :) it was addressed in the second commit of this pr, see: https://github.com/ManageIQ/manageiq/pull/5029/files#diff-bccee92afbf0d3d195b7bcf343b1b6c4R3453

blomquisg replied:
I'm a bit confused by this.

I can't tell if this was wrong to begin with, or what...  It's possible it was.  But, the [`miq_product_features.yml`](https://github.com/moolitayer/manageiq/blob/35949cd59972aa325d75907c2312162218c25686/db/fixtures/miq_product_features.yml#L3164-L3167) contained the `container_node_show` feature even though the original code (shown in my [comment above](https://github.com/ManageIQ/manageiq/pull/5029/files#r43759738)) called the feature `container_node`.

I don't know enough about our RBAC to know which one is right.

But, as you pointed out, your screenshot shows the right thing.  So, ¯\_(ツ)_/¯

@dclarizio, @h-kataria ... either of you have insight into this?  Should I just bury my head again?

h-kataria replied:
@blomquisg check in the original code was incorrect, it should have checked for container_node_show feature. container_node is a top parent feature id of container nodes related features.

simon3z commented:
@h-kataria @dclarizio can you review? Thanks!

chessbyte commented:
@moolitayer @simon3z Looks like legitimate test failures in Travis

moolitayer commented:
fixed tests, fixed container_node events handling conditional

moolitayer commented on app/views/container_group/show.html.haml:
```diff
@@ -1,6 +1,4 @@
-- if %w(containers container_projects).include?(@display)
-    = render :partial => "layouts/gtl", :locals => {:action_url => "show/#{@record.id}"}
-- elsif @display == 'container_services' && @showtype != "compare"
+- if %w(container_services containers).include?(@display)
```
@h-kataria Is it intentional that @showtype != "compare" was only for container_services
and not for containers? should I maintain it?

moolitayer commented:
rebased

simon3z commented:
@matthewd @blomquisg can you review?

simon3z commented:
LGTM :+1: cc @chessbyte

blomquisg commented:
@simon3z and @moolitayer, see [my comment](https://github.com/ManageIQ/manageiq/pull/5029#discussion_r43759738).  Let me know if I missed something there

moolitayer commented:
@blomquisg see [response](https://github.com/ManageIQ/manageiq/pull/5029/files#r43780514)

chessbyte commented:
@dclarizio @matthewd Please review

martinpovolny commented on app/helpers/application_helper.rb:
```diff
@@ -24,6 +24,42 @@ def miq_accordion_panel(title, condition, id, &block)
+      out = content_tag(:li) do
+        link_to("#{name}: #{ent.name}",
+                {:controller => table_name, :action => 'show', :id => ent.id.to_s},
+                :title => _("Show this #{record.class.name.demodulize.titleize}'s parent #{name}"))
```
wrong gettext usage here, please, see https://github.com/ManageIQ/manageiq/issues/3775

martinpovolny commented on app/helpers/application_helper.rb:
```diff
@@ -24,6 +24,42 @@ def miq_accordion_panel(title, condition, id, &block)
+        out = content_tag(:li) do
+          link_to("#{plural} (#{count})",
+                  {:action  => 'show', :id => @record, :display => table_name.to_s.pluralize},
+                  :title => _("Show #{plural}"))
```
wrong gettext usage

moolitayer commented:
follow #3775 for string interpolation inside a translated string (use % operator + parameter Hash)

simon3z commented on app/helpers/application_helper.rb:
```diff
@@ -24,6 +24,45 @@ def miq_accordion_panel(title, condition, id, &block)
+                {:controller => table_name, :action => 'show', :id => ent.id.to_s},
+                 :title => _("Show this %{entity_name}'s parent %{linked_entity_name}" %
+                            {:entity_name        => record.class.name.demodulize.titleize,
+                             :linked_entity_name => name})
```
@moolitayer I think that the internationalization is still wrong here. It should be:

    _("String with %{parameters}") % { ...parameters... }

and *not*:

    _("String with %{parameters}" % { ...parameters... })

Right @martinpovolny ?

martinpovolny replied:
Right. We need to get the string `"String with %{parameters}"` to get into the catalog. That is something that translators can work with,

moolitayer commented:
addressed comment

simon3z commented:
@blomquisg @chessbyte can you review/merge? Thanks!

simon3z commented:
> - Line 36, Col 18 - Style/AlignParameters - Align the parameters of a method call if they span more than one line.
>   - Line 39, Col 9 - Style/ClosingParenthesisIndentation - Align ) with (.
>   - Line 58, Col 27 - Style/ExtraSpacing - Unnecessary spacing detected.
>   - Line 59, Col 20 - Style/AlignParameters - Align the parameters of a method call if they span

@moolitayer I think these should be solvable.

zeari commented on app/helpers/application_helper.rb:
```diff
@@ -24,6 +24,45 @@ def miq_accordion_panel(title, condition, id, &block)
+    out
+  end
+
+  def multiple_relationship_link(record, table_name)
```
I think it would be better to add options such as a custom ```feature```.
For example, the feature for ```containers``` is "containers" and not ```container_show_list```
you can see that in ```container_summary_helper.rb``` the method for containers is ```textual_link(@record.containers, :feature => "containers")```

I would model this after ```textual_summary_helper.rb``` since that stabilized and this isnt so different functionally.

moolitayer commented:
@zeari the role was indeed containers but that was wrong and it was fixed in https://github.com/moolitayer/manageiq/commit/5ce2ffd49b955e14ee5a8e636cb6697a186c12aa 
we can add an optional argument if/when we need it.

rebased. fixed rubocop comments. keeping an eye out on rubocop

simon3z commented:
Ping  @blomquisg @h-kataria @chessbyte  LGTM :+1:

moolitayer commented:
> Ping @blomquisg @h-kataria @chessbyte LGTM

rebased. please review

moolitayer commented:
@miq-bot add_label backport/yes

moolitayer commented:
@blomquisg @h-kataria @chessbyte please review

chessbyte commented:
@matthewd @dclarizio @h-kataria @martinpovolny Please review as there is a bunch of UI refactoring in this PR.

chessbyte commented:
@Fryguy @jrafanie @bdunne break out the emojis for the net deletions of code. ;-)

matthewd commented:
Is this a bug fix, or a refactoring? :confused:

moolitayer commented:
> Is this a bug fix, or a refactoring?
Both. I went over all the container screens and fixed some inconsistencies we had there. 
#4950 and #4893 where fixed in the process as well as some other issues.

jrafanie commented on app/views/layouts/listnav/_container_group.html.haml:
```diff
@@ -33,45 +33,9 @@
-                {:controller => 'container_replicator', :action => 'show', :id => @record.container_replicator},
-                :title => _("Show %s %s") % [ui_lookup(:table => "container_replicator"),
-                  @record.container_replicator.name])
+        = single_relationship_link(@record, :ems_container, "ext_management_system")
```
Thank you @moolitayer logic in views are so hard to read... This is great.   Look at all of the deletes!  :heart_eyes:

bdunne commented:
Wow, lots of `if`'s removed :+1:

dclarizio commented:
@moolitayer this looks good and looks like we can reuse for other listnav areas.  Needs to rebase and if you tested and are satisfied with how it works, ready to merge.  Thx, Dan

moolitayer commented:
rebased and tested all screens again. verified both bug fixes. @jrafanie @bdunne @dclarizio thanks

moolitayer commented:
@matthewd why was the backport/yes label removed?

martinpovolny commented on app/helpers/application_helper.rb:
```diff
@@ -24,6 +24,44 @@ def miq_accordion_panel(title, condition, id, &block)
+    property_name ||= table_name
+    ent = record.send(property_name)
+    name = ui_lookup(:table => table_name.to_s)
+    if role_allows(:feature => "#{table_name}_show") && !ent.nil?
```
`ent.present?` maybe?

martinpovolny commented on app/helpers/application_helper.rb:
```diff
@@ -24,6 +24,44 @@ def miq_accordion_panel(title, condition, id, &block)
+        link_to("#{name}: #{ent.name}",
+                {:controller => table_name, :action => 'show', :id => ent.id.to_s},
+                :title       => _("Show this %{entity_name}'s parent %{linked_entity_name}") %
+                                {:entity_name        => record.class.name.demodulize.titleize,
```
not sure about this `record.class.name.demodulize.titleize` for the i18n, @mzazrivec: should we replace that with some `ui_lookup` call?

martinpovolny commented:
Great refactoring, saves a bunch of lines. Thx! 

We should surely expand that to more listnavs. And the function surely deserves a spec or 2 before we use it everywhere.

martinpovolny commented:
Any future refactoring/cleanup should compare and consider unification of this new helper method and previously existing `li_link_if_condition`
